runtime/near-vm-logic/src/tests/miscs.rs

```diff
@@ -876,70 +876,194 @@ fn test_ed25519_verify() {
         107, 108, 97, 100, 106, 102, 107, 108, 106, 97, 100, 115, 107,
     ];
 
-    let result = logic
-        .ed25519_verify(
-            signature.len() as _,
-            signature.as_ptr() as _,
-            message.len() as _,
-            message.as_ptr() as _,
-            public_key.len() as _,
-            public_key.as_ptr() as _,
-        )
-        .unwrap();
-
-    assert_eq!(result, 1);
-
-    assert_costs(map! {
-        ExtCosts::read_memory_byte: 128,
-        ExtCosts::read_memory_base: 3,
-        ExtCosts::ed25519_verify_base: 1,
-        ExtCosts::ed25519_verify_byte: 32,
-    });
+    let wrong_message_zero_bytes: [u8; 0] = [];
+
+    let scenarios = [
+        (
+            signature.len(),
+            signature.clone(),
+            message.len(),
+            message.as_slice(),
+            public_key.len(),
+            public_key.clone(),
+            Ok(1),
+        ),
+        (
+            signature.len(),
+            signature.clone(),
+            message.len(),
+            message.as_slice(),
+            public_key.len(),
+            forged_public_key.clone(),
+            Ok(0),
+        ),
+        (
+            signature.len(),
+            signature.clone(),
+            message.len(),
+            message.as_slice(),
+            public_key.len() - 1,
+            public_key.clone(),
+            Err(VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput {
+                msg: "invalid public key length".to_string(),
+            })),
+        ),
+        (
+            bad_signature.len(),
+            bad_signature.clone(),
+            message.len(),
+            message.as_slice(),
+            public_key.len(),
+            public_key.clone(),
+            Ok(0),
+        ),
+        (
+            signature.len() - 1,
+            signature.clone(),
+            message.len(),
+            message.as_slice(),
+            public_key.len(),
+            public_key.clone(),
+            Err(VMLogicError::HostError(HostError::Ed25519VerifyInvalidInput {
+                msg: "invalid signature length".to_string(),
+            })),
+        ),
+        (
+            forged_signature.len(),
+            forged_signature.clone(),
+            message.len(),
+            message.as_slice(),
+            public_key.len(),
+            public_key.clone(),
+            Ok(0),
+        ),
+        (
+            forged_signature.len(),
+            forged_signature.clone(),
+            0,
+            message.as_slice(),
+            public_key.len(),
+            public_key.clone(),
+            Ok(0),
+        ),
+        (
+            signature.len(),
+            signature.clone(),
+            message.len(),
+            wrong_message_zero_bytes.as_slice(), // note that here the message len is correct - TODO: ensure that this is the desired behavior
+            public_key.len(),
+            public_key.clone(),
+            Ok(0),
+        ),
```

Comment:
I think we should remove this case: it basically tests reading out-of-bounds.

That is, the calling contract certainly *can* write this kind of code, and the runtime would go and interperet whatever is in the heap as the message, but we can't detect this from within the runtime.